src/Mir/Intrinsics.hs

```diff
@@ -976,6 +994,46 @@ mirRef_arrayAsMirVectorIO sym _ (MirReference_Integer _ _) =
     addFailedAssertion sym $ GenericSimError $
         "attempted Array->MirVector conversion on the result of an integer-to-pointer cast"
 
+mirRef_eqIO :: IsSymInterface sym => sym ->
+    MirReference sym tp -> MirReference sym tp -> IO (RegValue sym BoolType)
+mirRef_eqIO sym (MirReference _ _) (MirReference _ _) =
+    -- TODO: implement an equality check for valid references
+    return $ falsePred sym
+mirRef_eqIO sym (MirReference_Integer _ i1) (MirReference_Integer _ i2) =
+    isEq sym i1 i2
+mirRef_eqIO sym _ _ =
+    -- All valid references are disjoint from all integer references.
+    return $ falsePred sym
+
+mirRef_offsetIO :: IsSymInterface sym => sym ->
+    TypeRepr tp -> MirReference sym tp -> RegValue sym IsizeType -> IO (MirReference sym tp)
+mirRef_offsetIO sym _tpr (MirReference root (Index_RefPath tpr path idx)) offset = do
+    idx' <- bvAdd sym idx offset
+    -- TODO: `offset` has a number of preconditions that we should check here:
```

Comment:
This seems like a key set of side conditions to check. Is that hard to do from this context?

Reply:
Not hard, I don't think, but a bit annoying to implement.  (Aside from the one about total offset, which requires size information that we don't export from mir-json yet.)  I don't think these checks are high priority for our current use case (that is, being able to evaluate parts of the standard library), though we'll definitely want to add the checks if we ever want to start verifying actual unsafe code.  I opened #21 to track this.